### ISSUE TITLE
Deadlock in DatadogStatsdLineBuilder

### ISSUE THREAD

**Issue by asvirsky** [micrometer-metrics/micrometer#2186]
We are using version 1.5.1, but based on my reading of the latest code, the problem is still there: map.computeIfAbsent is implicitly locking map key, while calling a method to create new entry. This method tries to acquire globally visible lock - "conventionTagsLock" - which may actually be owned by another thread _also_ polling all registered meters.

While deadlock is definitely a very serious issue, I mostly wonder why do we have more than one thread collecting and sending metric values? Is this by design, and is there a way to enforce max number of polling threads?

```
"parallel-7"
  waiting to lock java.lang.Object@a465b61
  which is held by "parallel-1"
"parallel-1"
  waiting to lock java.util.concurrent.ConcurrentHashMap$ReservationNode@5f8970dd
  which is held by "parallel-7"
```

```
"parallel-1" #58
   java.lang.Thread.State: BLOCKED
        at java.util.concurrent.ConcurrentHashMap.clear(ConcurrentHashMap.java:1193)
        - waiting to lock java.util.concurrent.ConcurrentHashMap$ReservationNode@5f8970dd
        at io.micrometer.statsd.internal.DatadogStatsdLineBuilder.updateIfNamingConventionChanged(DatadogStatsdLineBuilder.java:55)
        - locked on java.lang.Object@a465b61
        at io.micrometer.statsd.internal.DatadogStatsdLineBuilder.line(DatadogStatsdLineBuilder.java:45)
        at io.micrometer.statsd.internal.FlavorStatsdLineBuilder.count(FlavorStatsdLineBuilder.java:46)
        at io.micrometer.statsd.StatsdLineBuilder.count(StatsdLineBuilder.java:30)
        at io.micrometer.statsd.StatsdFunctionCounter.lambda$poll$0(StatsdFunctionCounter.java:46)
        at io.micrometer.statsd.StatsdFunctionCounter$$Lambda$626/617929212.apply()
        at java.util.concurrent.atomic.AtomicReference.updateAndGet(AtomicReference.java:179)
        at io.micrometer.statsd.StatsdFunctionCounter.poll(StatsdFunctionCounter.java:44)
        at io.micrometer.statsd.StatsdMeterRegistry.poll(StatsdMeterRegistry.java:175)
        at io.micrometer.statsd.StatsdMeterRegistry.lambda$startPolling$13(StatsdMeterRegistry.java:282)
        at io.micrometer.statsd.StatsdMeterRegistry$$Lambda$139/1446433477.accept()
        at io.micrometer.shaded.reactor.core.publisher.FluxDoOnEach$DoOnEachSubscriber.onNext(FluxDoOnEach.java:140)
        at io.micrometer.shaded.reactor.core.publisher.FluxInterval$IntervalRunnable.run(FluxInterval.java:123)
        at io.micrometer.shaded.reactor.core.scheduler.PeriodicWorkerTask.call(PeriodicWorkerTask.java:59)
        at io.micrometer.shaded.reactor.core.scheduler.PeriodicWorkerTask.run(PeriodicWorkerTask.java:73)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
```

```
"parallel-7" #412
   java.lang.Thread.State: BLOCKED
        at io.micrometer.statsd.internal.DatadogStatsdLineBuilder.ddTag(DatadogStatsdLineBuilder.java:82)
        - waiting to lock java.lang.Object@a465b61
        at io.micrometer.statsd.internal.DatadogStatsdLineBuilder$$Lambda$402/780589468.apply()
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1660)
        - locked on java.util.concurrent.ConcurrentHashMap$ReservationNode@5f8970dd
        at io.micrometer.statsd.internal.DatadogStatsdLineBuilder.tagsByStatistic(DatadogStatsdLineBuilder.java:78)
        at io.micrometer.statsd.internal.DatadogStatsdLineBuilder.line(DatadogStatsdLineBuilder.java:46)
        at io.micrometer.statsd.internal.FlavorStatsdLineBuilder.count(FlavorStatsdLineBuilder.java:46)
        at io.micrometer.statsd.StatsdLineBuilder.count(StatsdLineBuilder.java:30)
        at io.micrometer.statsd.StatsdFunctionCounter.lambda$poll$0(StatsdFunctionCounter.java:46)
        at io.micrometer.statsd.StatsdFunctionCounter$$Lambda$626/617929212.apply()
        at java.util.concurrent.atomic.AtomicReference.updateAndGet(AtomicReference.java:179)
        at io.micrometer.statsd.StatsdFunctionCounter.poll(StatsdFunctionCounter.java:44)
        at io.micrometer.statsd.StatsdMeterRegistry.poll(StatsdMeterRegistry.java:175)
        at io.micrometer.statsd.StatsdMeterRegistry.lambda$startPolling$13(StatsdMeterRegistry.java:282)
        at io.micrometer.statsd.StatsdMeterRegistry$$Lambda$139/1446433477.accept()
        at io.micrometer.shaded.reactor.core.publisher.FluxDoOnEach$DoOnEachSubscriber.onNext(FluxDoOnEach.java:140)
        at io.micrometer.shaded.reactor.core.publisher.FluxInterval$IntervalRunnable.run(FluxInterval.java:123)
        at io.micrometer.shaded.reactor.core.scheduler.PeriodicWorkerTask.call(PeriodicWorkerTask.java:59)
        at io.micrometer.shaded.reactor.core.scheduler.PeriodicWorkerTask.run(PeriodicWorkerTask.java:73)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)

```

**Comment by izeye** [comment]
@asvirsky Thanks for the report, but this seems to be a duplicate of https://github.com/micrometer-metrics/micrometer/issues/2100 that has been fixed in 1.5.2. Could you try 1.5.2? If it still happens, please share its thread dump.

**Comment by asvirsky** [comment]
@izeye yes, indeed, it looks pretty similar - thanks for pointing it out. Will upgrade to 1.5.2 and hope for the best...
I don't really know how to reproduce the problem. Somehow 2 servers got into this state, while the other 50 seem to be OK. 

I also had couple of questions about large number of threads - "parallel-1", "parallel-2", ... "parallel-24", etc. Right after launching jvm only one of them is created, but shortly after new ones are starting to pop out. It doesn't seem like they ever expire (die), although i'm not 100% sure. I was never able to catch them doing anything, until we ran into deadlock, when all of them got stuck almost like couple of deadlocked threads (but w/out actual *dead*lock). Stack trace suggests that polling process was scheduled, rather than  triggered by meter update.

```
"parallel-3" #403
   java.lang.Thread.State: BLOCKED
        at java.util.concurrent.ConcurrentHashMap.computeIfAbsent(ConcurrentHashMap.java:1674)
        - waiting to lock java.util.concurrent.ConcurrentHashMap$ReservationNode@5f8970dd
        at io.micrometer.statsd.internal.DatadogStatsdLineBuilder.tagsByStatistic(DatadogStatsdLineBuilder.java:78)
        at io.micrometer.statsd.internal.DatadogStatsdLineBuilder.line(DatadogStatsdLineBuilder.java:46)
        at io.micrometer.statsd.internal.FlavorStatsdLineBuilder.count(FlavorStatsdLineBuilder.java:46)
        at io.micrometer.statsd.StatsdLineBuilder.count(StatsdLineBuilder.java:30)
        at io.micrometer.statsd.StatsdFunctionCounter.lambda$poll$0(StatsdFunctionCounter.java:46)
        at io.micrometer.statsd.StatsdFunctionCounter$$Lambda$626/617929212.apply()
        at java.util.concurrent.atomic.AtomicReference.updateAndGet(AtomicReference.java:179)
        at io.micrometer.statsd.StatsdFunctionCounter.poll(StatsdFunctionCounter.java:44)
        at io.micrometer.statsd.StatsdMeterRegistry.poll(StatsdMeterRegistry.java:175)
        at io.micrometer.statsd.StatsdMeterRegistry.lambda$startPolling$13(StatsdMeterRegistry.java:282)
        at io.micrometer.statsd.StatsdMeterRegistry$$Lambda$139/1446433477.accept()
        at io.micrometer.shaded.reactor.core.publisher.FluxDoOnEach$DoOnEachSubscriber.onNext(FluxDoOnEach.java:140)
        at io.micrometer.shaded.reactor.core.publisher.FluxInterval$IntervalRunnable.run(FluxInterval.java:123)
        at io.micrometer.shaded.reactor.core.scheduler.PeriodicWorkerTask.call(PeriodicWorkerTask.java:59)
        at io.micrometer.shaded.reactor.core.scheduler.PeriodicWorkerTask.run(PeriodicWorkerTask.java:73)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

```

I have 2 related questions:

1. is this really intended behavior (using multiple threads to poll) or is this an invalid state, caused by the same conditions that lead to  deadlock too? It is not hard to imagine how multiple threads would corrupt collected stats
1. are there any "secret" configuration settings I can use to adjust this logic, specifically how can change the maximum number of threads?

Would really appreciate if you or anyone else could answer.

Thank you!

**Comment by izeye** [comment]
> is this really intended behavior (using multiple threads to poll) or is this an invalid state, caused by the same conditions that lead to deadlock too? It is not hard to imagine how multiple threads would corrupt collected stats

@asvirsky It depends on `Schedulers.parallel()` from Reactor Netty for execution, and it creates workers up to `Runtime.getRuntime().availableProcessors()` unless `reactor.schedulers.defaultPoolSize` system property is specified.

> are there any "secret" configuration settings I can use to adjust this logic, specifically how can change the maximum number of threads?

Micrometer shades Reactor Netty, so you need to set `io.micrometer.shaded.reactor.schedulers.defaultPoolSize` system property instead, for example as follows:

```
-Dio.micrometer.shaded.reactor.schedulers.defaultPoolSize=6
```

**Comment by asvirsky** [comment]
We upgraded to 1.5.2 and indeed it does fix the deadlock problem...

Unfortunately, it comes with even worse issue: shortly after start udp-nio-1 thread is going into endless DatagramChannel open/close loop, consuming 100% of CPU. It doesn't happen on all servers, but  where it does happen it seems very consistent.  
On a server with 40 cores it wasn't instantly obvious that one core is busy all the time. But heap usage pattern was more than suspicious: every 2 minutes 30+Gb are allocated, and then released (causing 1 second GC pause):

![image](https://user-images.githubusercontent.com/3719307/87896859-5ac5a580-c9fe-11ea-85fb-99d188ca2808.png)


Stack traces:
(1)
![image](https://user-images.githubusercontent.com/3719307/87896305-abd49a00-c9fc-11ea-8696-79df2f9f36fe.png)


(2) 
![image](https://user-images.githubusercontent.com/3719307/87896337-c870d200-c9fc-11ea-9b94-5859fe00f5a1.png)

(3)
![image](https://user-images.githubusercontent.com/3719307/87896355-d58dc100-c9fc-11ea-88f0-c57b61a17922.png)

**Comment by shakuzen** [comment]
> We upgraded to 1.5.2 and indeed it does fix the deadlock problem...
> 
> Unfortunately, it comes with even worse issue: shortly after start udp-nio-1 thread is going into endless DatagramChannel open/close loop, consuming 100% of CPU. It doesn't happen on all servers, but where it does happen it seems very consistent.

@asvirsky that should be fixed with #2177. It sounds like we can close this issue as a duplicate of #2100.
